feat(figma-extractor): implement figma Token extractor (#3357)
AmadeusITGroup/otter: docs/figma-extractor/figma-extractor-setup.md

# Figma Extractor

The Figma Extractor tool allows the extraction of [Design Tokens](https://www.designtokens.org/tr/drafts/format/) from a single Figma file or a Figma project.\
The extractor will help you to:

- Manage the version of your Figma File
- Generate Design Token specification from a given Figma File (or Figma Project).
- Generating an NPM artifact containing your Design Token (and then benefit of Renovate to automate the upgrade and CSS regeneration)

> [!NOTE]
> More details regarding the command is available on [@ama-styling/figma-extractor readme](https://www.npmjs.com/package/@ama-styling/figma-extractor).

> ## Guidelines

Reviewer: Maybe replace the Guidelines here as a warning or info section as it does not make sense as a standalone section

As Figma is currently missing features such as versioning and proper release management, a few assumptions and guidelines need to be taken by the extractor to avoid storing information outside from Figma.

## Extractor usage

The list of the commands and options supported by the Figma Extract is available on [@ama-styling/figma-extractor readme](https://www.npmjs.com/package/@ama-styling/).\
This section focuses on the data to provide the CLI and in which context it can be used.

### Access Permissions

To access the Figma File(s), a [Figma Access Token](https://help.figma.com/hc/en-us/articles/8085703771159-Manage-personal-access-tokens) needs to be provided to the CLI via the `--accessToken` option or via the environment variable `FIGMA_TOKEN`.

The provided token should be configured with the following **read** permissions:

- **File content**: to access to the file components (including Effects, gradients, shadows, etc...)

> - **Library assets**: to access to external referred components

Reviewer: ```suggestion
- **Library assets**: to access to external referred components.
```

> - **Library content**: to access to library published components

Reviewer: ```suggestion
- **Library content**: to access to library published components.
```

- **File versions**: to determine the latest version set to the tokens, see [versioning](#versioning).
- **Variables**: to retrieve the list of variables to convert to Design Token.
- ***Projects*** (`extract-project` command only): retrieve the list of files withing a project to determine the latest one base on version, see [versioning](#versioning).

### Usage as NPM Package

> The option `--generatePackage` of the CLI generates a [NPM Package](https://docs.npmjs.com/about-packages-and-modules). The package name can be configured thanks to the `--name` option. Its version is determined by the Figma file information (see [versioning](#versioning)).

Reviewer: ```suggestion
The option `--generatePackage` of the CLI generates a [NPM Package](https://docs.npmjs.com/about-packages-and-modules) instead of a Design Token file. The package name can be configured thanks to the `--name` option. Its version is determined by the Figma file information (see [versioning](#versioning)).
```

>
> - Benefit from version update through bot such as [Renovate](https://docs.renovatebot.com/).
> - Use it as the source for your [Style Dictionary](https://styledictionary.com/) to generate your css.

You can then use the generated NPM Artifact as dependency of your application/library to generate your CSS following the next example:

```json5
// App/Lib package.json

{
  "name": "my-app",
  "devDependencies": {
    "@my-project/design-tokens": "*",
    "style-dictionary": "*"
  }
}
```

```javascript
// config.mjs

import { dirname, join } from 'node:path';

/** @type {import('style-dictionary').Config} */
export default {
  usesDtcg: true, // Use Design Token Standard format

  // List the files from your Design Tokens NPM package
  source: join(dirname(require.resolve('@my-project/design-tokens')), '*.json'),

  platforms: {
    // your configuration
  }
};

```

> [!TIP]
> See [@ama-styling/style-dictionary](https://www.npmjs.com/package/@ama-styling/style-dictionar) to setup a `config.mjs` file

### Versioning

Figma does not provide an explicit versioning mechanism for a Figma File.\

> To make sure to have the proper information regarding the update a Design Token, it is important to apply manually a versioning following the [Semantic Versioning](https://semver.org/) standard.

Reviewer: ```suggestion
You can convey meaning to your Design Token updates (such as the introduction of a breaking feature, a non-breaking feature, a fix) by following the [Semantic Versioning](https://semver.org/) standard.

This way, the increase of a major will inform developers that the design of potential removal of tokens and migration efforts to apply the new design.

Let's see how to do it.
```

#### Label in File History

The [History mechanism](https://help.figma.com/hc/en-us/articles/360038006754-View-a-file-s-version-history) of Figma allows the user to identify a specific revision thanks to a **Label**.
If you want the Figma Extractor to properly tag your Design Tokens and establish a consistent versioning history, ensure that the History labels follow the SemVer standard.

#### File Versioning

> To simplified the segmentation of **Major release version**, it can be interesting to have mulliple Figma File for each major versions.

Reviewer: ```suggestion
To simplified the segmentation of **Major release version**, it can be interesting to have multiple Figma Files for each major versions.
```

But honestly I am not sure what it means 😅

> To allow it the Figma Extractor expose the `extract-project` which will consider a whole Figma Project instead of a single Figma File.\

Reviewer: ```suggestion
If your project follows this architecture, running the extractor with the `extract-project` option will extract the whole Figma Project instead of a single Figma File.\
```

# Figma Extractor

> The Figma Extractor tool allows the extraction of [Design Tokens](https://www.designtokens.org/tr/drafts/format/) from a single Figma file or a Figma project.

Reviewer: What is documentation trying to achieve?

> [!NOTE]
> More details regarding the command is available on [@ama-styling/figma-extractor readme](https://www.npmjs.com/package/@ama-styling/figma-extractor).

## Guidelines

To complete Figma currently missing feature, few assumptions and guidelines has been taken by the extractor to let data to be driven by Figma.

> ### Versioning solution

Reviewer: ```suggestion
### Versioning
```

Figma does not provide explicit versioning mechanism for a Figma File.\

> To make sure to have the proper information regarding the update a Design Token, it is important to apply manually a versioning following the [Semantioc Versioning](https://semver.org/) standard.

Reviewer: What proper information do we expect? I am not sure what you mean here.
What information are we trying to convey here?

#### Label in File History

> Figma offers a [History mechanism](https://help.figma.com/hc/en-us/articles/360038006754-View-a-file-s-version-history) allowing to defined a **Label** to a specific revision.

Reviewer: ```suggestion
The [History mechanism](https://help.figma.com/hc/en-us/articles/360038006754-View-a-file-s-version-history) of Figma allows the user to identify a specific revision thanks to a **Label**.
```

> In manner to get a propper version list determined by the Figma Extractor, the history labels corresponding to a SemVer version can be applied.

Reviewer: ```suggestion
If you want the Figma Extractor to properly tag your Design Tokens and establish a consistent versioning history, ensure that the History labels follow the SemVer standard.
```

#### File Versioning

> To simplified the segmentation of **Major release version**, it can be interesting to have mulliple Figma File for each major versions.

Reviewer: What is the segmentation of the Major release version? What is it for?

To allow it the Figma Extractor expose the `extract-project` which will consider a whole Figma Project instead of a single Figma File.\
Thanks to the option `--filenamePattern` it is possible to help the Figma Extractor to discover the file version to determine the latest one of the Figma Project.

> [!TIP]
> A limit of the considered version can be specified via the option `--versionRange`.

## Extractor usage

The list of the commands and options supported by the Figma Extract is available on [@ama-styling/figma-extractor readme](https://www.npmjs.com/package/@ama-styling/).\
This session of the documentation will focus on the information to provide to the CLI and the context where it can be used.

### Access Permissions

To properly access to the Figma File(s), a [Figma Access Token](https://help.figma.com/hc/en-us/articles/8085703771159-Manage-personal-access-tokens) need to be provided to the CLI via the `--accessToken` (or the environment variable `FIGMA_TOKEN`).

> The provided token should have the following **read** permissions:

Reviewer: ```suggestion
The provided token should be configured with the following **read** permissions:
```

- **File content**: to access to the file components (including Effects, gradients, shadows, etc...)
- **Library assets**: to access to external referred components
- **Library content**: to access to library published components
- **File versions**: to determine the latest version set to the tokens, see [versioning](#versioning-solution).
- **Variables**: to retrieve the list of variables to convert to Design Token.
- ***Projects*** (`extract-project` command only): retrieve the list of files withing a project to determine the latest one base on version, see [versioning](#versioning-solution).

> ### Usage as NPM Package

Reviewer: This section should probably be introduced before the versioning one as this is the purpose of the package.
